File: sdks/ts/memory/src/ratelimit/adaptive.test.ts
// SPDX-License-Identifier: Apache-2.0

/**
 * Unit tests for the adaptive rate limiter wrapper.
 */

import { afterEach, describe, expect, it } from 'vitest'
import type { RateLimiter } from './types.js'
import { createAdaptiveRateLimiter } from './adaptive.js'
import { createTokenBucket } from './token-bucket.js'

describe('createAdaptiveRateLimiter', () => {
  const limitersTeardown: RateLimiter[] = []

  afterEach(async () => {
    for (const lim of limitersTeardown.splice(0)) {
      await lim.close()
    }
  })

  const makeAdaptive = (
    bucketOpts: Partial<Parameters<typeof createTokenBucket>[0]> = {},
    adaptiveOpts: Partial<Parameters<typeof createAdaptiveRateLimiter>[0]> = {},
  ): RateLimiter => {
    const bucket = createTokenBucket({
      maxTokens: 100,
      refillRatePerSecond: 50,
      tenantId: 'test',
      ...bucketOpts,
    })
    const lim = createAdaptiveRateLimiter({
      bucket,
      minRefillRate: 1,
      maxRefillRate: 100,
      recoveryFactor: 1.5,
      ...adaptiveOpts,
    })
    limitersTeardown.push(lim)
    return lim
  }

  it('reduces refill rate when remaining is low', () => {
    const lim = makeAdaptive()

    // remaining=5 < threshold (100/4=25) -> throttle.
    lim.updateFromHeaders({
      remaining: 5,
      limit: 100,
    })

    const m = lim.metrics()
    expect(m.refillRatePerSecond).toBeLessThan(50)
  })

  it('recovers rate when remaining is above threshold', () => {
    const lim = makeAdaptive({ refillRatePerSecond: 10 })

    const before = lim.metrics().refillRatePerSecond

    lim.updateFromHeaders({
      remaining: 80,
      limit: 100,
    })

    const m = lim.metrics()
    expect(m.refillRatePerSecond).toBeGreaterThan(before)
  })

  it('rate never drops below minRefillRate', () => {
    const lim = makeAdaptive(
      { refillRatePerSecond: 2 },
      { minRefillRate: 5 },
    )

    // Repeatedly throttle.
    for (let i = 0; i < 20; i++) {
      lim.updateFromHeaders({ remaining: 1, limit: 100 })
    }

    const m = lim.metrics()
    expect(m.refillRatePerSecond).toBeGreaterThanOrEqual(5)
  })

  it('rate never exceeds maxRefillRate', () => {
    const lim = makeAdaptive(
      { refillRatePerSecond: 80 },
      { maxRefillRate: 100, recoveryFactor: 2.0 },
    )

    // Repeatedly recover.
    for (let i = 0; i < 20; i++) {
      lim.updateFromHeaders({ remaining: 90, limit: 100 })
    }

    const m = lim.metrics()
    expect(m.refillRatePerSecond).toBeLessThanOrEqual(100)
  })

  it('delegates retry-after to the underlying bucket', () => {
    const bucket = createTokenBucket({
      maxTokens: 10,
      refillRatePerSecond: 50,
      tenantId: 'test',
    })
    const lim = createAdaptiveRateLimiter({
      bucket,
      minRefillRate: 1,
      maxRefillRate: 100,
    })
    limitersTeardown.push(lim)

    lim.updateFromHeaders({ retryAfter: 0.05 })

    const bm = bucket.metrics()
    expect(bm.refillRatePerSecond).toBe(0)
  })

  it('acquire delegates to underlying bucket', async () => {
    const lim = makeAdaptive()
    const tok = await lim.acquire(1)
    expect(tok).toBeDefined()
    tok.release()
  })

  it('tryAcquire delegates to underlying bucket', () => {
    const lim = makeAdaptive()
    const tok = lim.tryAcquire(1)
    expect(tok).toBeDefined()
    tok!.release()
  })

  it('uses default options when not specified', () => {
    const bucket = createTokenBucket({
      maxTokens: 10,
      refillRatePerSecond: 10,
      tenantId: 'test',
    })
    const lim = createAdaptiveRateLimiter({ bucket })
    limitersTeardown.push(lim)

    // Should not throw.
    const m = lim.metrics()
    expect(m.refillRatePerSecond).toBe(10)
  })
})

File: go/ratelimit/adaptive_test.go
// SPDX-License-Identifier: Apache-2.0

package ratelimit

import (
	"context"
	"testing"
	"time"
)

func TestAdaptive_ThrottlesOnLowRemaining(t *testing.T) {
	bucket := NewBucket(BucketOptions{
		MaxTokens:        100,
		RefillRatePerSec: 50,
		TenantID:         "t1",
	})
	a := NewAdaptive(AdaptiveOptions{
		Bucket:         bucket,
		MinRefillRate:  1,
		MaxRefillRate:  100,
		RecoveryFactor: 1.5,
	})
	defer a.Close()

	// remaining=5 < threshold (100/4=25) -> throttle.
	a.UpdateFromHeaders(Headers{
		Remaining: 5,
		Limit:     100,
	})

	m := a.Metrics()
	if m.RefillRatePerSec >= 50 {
		t.Fatalf("expected reduced rate after throttle, got %f", m.RefillRatePerSec)
	}
}

func TestAdaptive_RecoversWhenRemainingAboveThreshold(t *testing.T) {
	bucket := NewBucket(BucketOptions{
		MaxTokens:        100,
		RefillRatePerSec: 10,
		TenantID:         "t1",
	})
	a := NewAdaptive(AdaptiveOptions{
		Bucket:         bucket,
		MinRefillRate:  1,
		MaxRefillRate:  100,
		RecoveryFactor: 2.0,
	})
	defer a.Close()

	initialRate := a.Metrics().RefillRatePerSec

	// Remaining above threshold -> recovery.
	a.UpdateFromHeaders(Headers{
		Remaining: 80,
		Limit:     100,
	})

	m := a.Metrics()
	if m.RefillRatePerSec <= initialRate {
		t.Fatalf("expected increased rate after recovery, got %f (was %f)", m.RefillRatePerSec, initialRate)
	}
}

func TestAdaptive_RateNeverDropsBelowMin(t *testing.T) {
	bucket := NewBucket(BucketOptions{
		MaxTokens:        100,
		RefillRatePerSec: 2,
		TenantID:         "t1",
	})
	a := NewAdaptive(AdaptiveOptions{
		Bucket:         bucket,
		MinRefillRate:  5,
		MaxRefillRate:  100,
		RecoveryFactor: 1.5,
	})
	defer a.Close()

	// Repeatedly throttle to drive rate down.
	for range 20 {
		a.UpdateFromHeaders(Headers{
			Remaining: 1,
			Limit:     100,
		})
	}

	m := a.Metrics()
	if m.RefillRatePerSec < 5 {
		t.Fatalf("rate %f dropped below min 5", m.RefillRatePerSec)
	}
}

func TestAdaptive_RateNeverExceedsMax(t *testing.T) {
	bucket := NewBucket(BucketOptions{
		MaxTokens:        100,
		RefillRatePerSec: 80,
		TenantID:         "t1",
	})
	a := NewAdaptive(AdaptiveOptions{
		Bucket:         bucket,
		MinRefillRate:  1,
		MaxRefillRate:  100,
		RecoveryFactor: 2.0,
	})
	defer a.Close()

	// Repeatedly recover to drive rate up.
	for range 20 {
		a.UpdateFromHeaders(Headers{
			Remaining: 90,
			Limit:     100,
		})
	}

	m := a.Metrics()
	if m.RefillRatePerSec > 100 {
		t.Fatalf("rate %f exceeded max 100", m.RefillRatePerSec)
	}
}

func TestAdaptive_DelegatesRetryAfterToBucket(t *testing.T) {
	bucket := NewBucket(BucketOptions{
		MaxTokens:        10,
		RefillRatePerSec: 50,
		TenantID:         "t1",
	})
	a := NewAdaptive(AdaptiveOptions{
		Bucket:        bucket,
		MinRefillRate: 1,
		MaxRefillRate: 100,
	})
	defer a.Close()

	a.UpdateFromHeaders(Headers{
		RetryAfter: 50 * time.Millisecond,
	})

	// The bucket should be paused.
	bm := bucket.Metrics()
	if bm.RefillRatePerSec != 0 {
		t.Fatalf("expected bucket rate 0 during retry-after, got %f", bm.RefillRatePerSec)
	}
}

func TestAdaptive_AcquireWorks(t *testing.T) {
	bucket := NewBucket(BucketOptions{
		MaxTokens:        10,
		RefillRatePerSec: 100,
		TenantID:         "t1",
	})
	a := NewAdaptive(AdaptiveOptions{
		Bucket: bucket,
	})
	defer a.Close()

	tok, err := a.Acquire(context.Background(), 1)
	if err != nil {
		t.Fatalf("acquire failed: %v", err)
	}
	tok.Release()
}

func TestAdaptive_TryAcquireWorks(t *testing.T) {
	bucket := NewBucket(BucketOptions{
		MaxTokens:        10,
		RefillRatePerSec: 100,
		TenantID:         "t1",
	})
	a := NewAdaptive(AdaptiveOptions{
		Bucket: bucket,
	})
	defer a.Close()

	tok, ok := a.TryAcquire(1)
	if !ok {
		t.Fatalf("tryAcquire should succeed with tokens available")
	}
	tok.Release()
}

func TestAdaptive_DefaultOptions(t *testing.T) {
	bucket := NewBucket(BucketOptions{
		MaxTokens:        10,
		RefillRatePerSec: 10,
		TenantID:         "t1",
	})
	// All options at zero -> defaults should be applied.
	a := NewAdaptive(AdaptiveOptions{
		Bucket: bucket,
	})
	defer a.Close()

	// Verify it works without panicking.
	tok, err := a.Acquire(context.Background(), 1)
	if err != nil {
		t.Fatalf("acquire failed: %v", err)
	}
	tok.Release()
}

func TestAdaptive_ThrottleAppliesRateToBucket(t *testing.T) {
	bucket := NewBucket(BucketOptions{
		MaxTokens:        100,
		RefillRatePerSec: 50,
		TenantID:         "t1",
	})
	a := NewAdaptive(AdaptiveOptions{
		Bucket:         bucket,
		MinRefillRate:  1,
		MaxRefillRate:  100,
		RecoveryFactor: 1.5,
	})
	defer a.Close()

	// Throttle via adaptive.
	a.UpdateFromHeaders(Headers{
		Remaining: 5,
		Limit:     100,
	})

	// The adaptive rate should be applied to the underlying bucket.
	adaptiveRate := a.Metrics().RefillRatePerSec
	bucketRate := bucket.Metrics().RefillRatePerSec
	if adaptiveRate != bucketRate {
		t.Fatalf("adaptive rate %f should equal bucket rate %f", adaptiveRate, bucketRate)
	}
}

func TestAdaptive_NoDoubleThrottle(t *testing.T) {
	bucket := NewBucket(BucketOptions{
		MaxTokens:        100,
		RefillRatePerSec: 50,
		TenantID:         "t1",
	})
	a := NewAdaptive(AdaptiveOptions{
		Bucket:         bucket,
		MinRefillRate:  1,
		MaxRefillRate:  100,
		RecoveryFactor: 1.5,
	})
	defer a.Close()

	// The adaptive layer throttles based on remaining < threshold.
	a.UpdateFromHeaders(Headers{
		Remaining: 5,
		Limit:     100,
	})

	adaptiveRate := a.Metrics().RefillRatePerSec

	// The bucket should have the same rate as adaptive (no independent
	// back-off was applied on top).
	bucketRate := bucket.Metrics().RefillRatePerSec
	if adaptiveRate != bucketRate {
		t.Fatalf("double-throttle detected: adaptive=%f, bucket=%f", adaptiveRate, bucketRate)
	}
}

func TestAdaptive_SetRefillRate(t *testing.T) {
	bucket := NewBucket(BucketOptions{
		MaxTokens:        10,
		RefillRatePerSec: 50,
		TenantID:         "t1",
	})
	a := NewAdaptive(AdaptiveOptions{
		Bucket: bucket,
	})
	defer a.Close()

	a.SetRefillRate(25)
	m := a.Metrics()
	if m.RefillRatePerSec != 25 {
		t.Fatalf("expected 25, got %f", m.RefillRatePerSec)
	}
	bm := bucket.Metrics()
	if bm.RefillRatePerSec != 25 {
		t.Fatalf("expected bucket rate 25, got %f", bm.RefillRatePerSec)
	}
}
